test/helpers/integration.rb

```diff
@@ -139,56 +146,53 @@ def wait_for_server_to_boot(log: false)
     wait_for_server_to_include 'Ctrl-C', log: log
   end
 
-  # Returns true if and when server log includes str.
-  # Will timeout or raise an error otherwise
-  def wait_for_server_to_include(str, log: false)
-    sleep 0.05 until @server.is_a?(IO)
-    retry_cntr = 0
-    begin
-      @server.wait_readable 1
-      if log
-        puts "Waiting for '#{str}'"
-        begin
-          line = @server&.gets
-          puts line if !line&.strip.empty?
-        end until line&.include?(str)
-      else
-        true until (@server.gets || '').include?(str)
-      end
-    rescue Errno::EBADF, Errno::ECONNREFUSED, Errno::ECONNRESET, IOError => e
-      retry_cntr += 1
-      raise e if retry_cntr > 10
-      sleep 0.1
-      retry
-    end
+  # Returns true if and when server log includes str.  Will timeout otherwise.
+  def wait_for_server_to_include(str, timeout: LOG_TIMEOUT, log: false)
+    time_timeout = Process.clock_gettime(Process::CLOCK_MONOTONIC) + timeout
+    line = ''
+
+    puts "\n——— #{full_name} waiting for '#{str}'" if log
+    line = server_gets(str, time_timeout, log: log) until line&.include?(str)
     true
   end
 
   # Returns line if and when server log matches re, unless idx is specified,
-  # then returns regex match.
-  # Will timeout or raise an error otherwise
-  def wait_for_server_to_match(re, idx = nil, log: false)
-    sleep 0.05 until @server.is_a?(IO)
-    retry_cntr = 0
-    line = nil
+  # then returns regex match.  Will timeout otherwise.
+  def wait_for_server_to_match(re, idx = nil, timeout: LOG_TIMEOUT, log: false)
+    time_timeout = Process.clock_gettime(Process::CLOCK_MONOTONIC) + timeout
+    line = ''
+
+    puts "\n——— #{full_name} waiting for '#{re.inspect}'" if log
+    line = server_gets(re, time_timeout, log: log) until line&.match?(re)
+    idx ? line[re, idx] : line
+  end
+
+  def server_gets(match_obj, time_timeout, log: false)
+    error_retries = 0
+    line = ''
+
+    sleep 0.05 unless @server.is_a?(IO) or Process.clock_gettime(Process::CLOCK_MONOTONIC) > time_timeout
+
+    raise Minitest::Assertion,  "@server is not an IO" unless @server.is_a?(IO)
+    if Process.clock_gettime(Process::CLOCK_MONOTONIC) > time_timeout
+      raise Minitest::Assertion, "Timeout waiting for server to log #{match_obj.inspect}"
+    end
+
     begin
-      @server.wait_readable 1
-      if log
-        puts "Waiting for '#{re.inspect}'"
-        begin
-          line = @server&.gets
-          puts line if !line&.strip.empty?
-        end until line&.match?(re)
-      else
-        true until (line = @server.gets || '').match?(re)
+      if @server.wait_readable(LOG_WAIT_READ) and line = @server&.gets
+        @server_log << line
+        puts "    #{line}" if log
       end
-    rescue Errno::EBADF, Errno::ECONNREFUSED, Errno::ECONNRESET, IOError => e
-      retry_cntr += 1
-      raise e if retry_cntr > 10
-      sleep 0.1
+    rescue Exception => e
```

Comment:
Not enough with `StandardError` here?

Reply:
Ok, I'll change.  IIRC, sometimes servers won't start in CI, and I can't recall what error is raised.